README.md
# [中文](https://www.jianshu.com/p/031a36e99196)

# flutter_tableview
[![pub package](https://img.shields.io/badge/pub-v1.0.1-orange.svg)](https://github.com/chenfanfang/flutter_tableView)

A flutter widget like iOS UITableview. let listView with section header and each section header will hover at the top.

## Installing:

Add the following to your `pubspec.yaml` file:

    dependencies:
      flutter_tableView: ^1.0.1

## How to use

```dart
class SimpleDemoPageBody extends StatefulWidget {
  @override
  _SimpleDemoPageBodyState createState() => _SimpleDemoPageBodyState();
}

class _SimpleDemoPageBodyState extends State<SimpleDemoPageBody> {
  // How many section.
  int sectionCount = 3;

  // Get row count.
  int _rowCountAtSection(int section) {
    if (section == 0) {
      return 5;
    } else if (section == 1) {
      return 10;
    } else {
      return 20;
    }
  }

  // Section header widget builder.
  Widget _sectionHeaderBuilder(BuildContext context, int section) {
    return InkWell(
      onTap: () {
        print('click section header. -> section:$section');
      },
      child: Container(
        alignment: Alignment.centerLeft,
        padding: EdgeInsets.only(left: 16.0),
        color: Color.fromRGBO(220, 220, 220, 1),
        height: 100,
        child: Text('I am section header -> section:$section'),
      ),
    );
  }

  // cell item widget builder.
  Widget _cellBuilder(BuildContext context, int section, int row) {
    return InkWell(
      onTap: () {
        print('click cell item. -> section:$section row:$row');
      },
      child: Container(
        padding: EdgeInsets.only(left: 16.0),
        alignment: Alignment.centerLeft,
        decoration: BoxDecoration(
            border: Border(
                bottom: BorderSide(
          color: Color.fromRGBO(240, 240, 240, 1),
        ))),
        height: 50.0,
        child: Text('I am cell -> section:$section  row$row'),
      ),
    );
  }

  // Each section header height;
  double _sectionHeaderHeight(BuildContext context, int section) {
    return 50.0;
  }

  // Each cell item widget height.
  double _cellHeight(BuildContext context, int section, int row) {
    return 50.0;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      //FlutterTableView
      child: FlutterTableView(
        sectionCount: sectionCount,
        rowCountAtSection: _rowCountAtSection,
        sectionHeaderBuilder: _sectionHeaderBuilder,
        cellBuilder: _cellBuilder,
        sectionHeaderHeight: _sectionHeaderHeight,
        cellHeight: _cellHeight,
      ),
    );
  }
}
```


![demo_picture](https://raw.githubusercontent.com/chenfanfang/flutter_tableView/master/example/demo_picture/simpleDemo.png)

gif:<br/>
![simple_demo.gif](https://upload-images.jianshu.io/upload_images/1594675-b85a99f485c5a25f.gif?imageMogr2/auto-orient/strip)




##### If you want to wrap listView with other widget (such as [flutter_easyrefresh](https://github.com/xuelongqy/flutter_easyrefresh))



```dart
FlutterTableView(
        sectionCount: this.dataSourceList.length,
        rowCountAtSection: _rowCountAtSection,
        sectionHeaderBuilder: _sectionHeaderBuilder,
        cellBuilder: _cellBuilder,
        sectionHeaderHeight: _sectionHeaderHeight,
        cellHeight: _cellHeight,
        listViewFatherWidgetBuilder: (BuildContext context, Widget listView) {
          return EasyRefresh(
            key: _easyRefreshKey,
            limitScroll: true,
            refreshHeader: MaterialHeader(key: _headerKey),
            refreshFooter: MaterialFooter(key: _footerKey),
            onRefresh: () async {},
            loadMore: () async {},
            child: listView,
          );
        },
      ),
      
// detail usage please download demo
```

gif:<br/>
![wrap_refresh_demo.gif](https://upload-images.jianshu.io/upload_images/1594675-842791f8f08ca4ab.gif)

## License
MIT License

Copyright (c) 2019 chenfanfang

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
## [1.0.1] - 2019-06-17

* Edit README.MD


## [1.0.0] - 2019-06-15

* flutter_tableview first version published.


test/flutter_tableview_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_tableview/flutter_tableview.dart';

void main() {

}


example/lib/main.dart
//flutter
import 'package:flutter/material.dart';
//tool
import 'demo_tool/route_tool.dart';
//page
import 'demo_page/simple_demo_page.dart';
import 'demo_page/wrap_refresh_demo_page.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    initScreen(context: context);
    return Scaffold(
      appBar: AppBar(
        title: Text('FlutterTableView'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FlatButton(
              onPressed: () {
                push(context, SimpleDemoPage());
              },
              child: Text('simple demo'),
            ),
            FlatButton(

              onPressed: () {
                push(context, WrapRefreshDemoPage());

              },
              child: Text('wrap refresh demo'),
            ),
          ],
        ),
      ),
    );
  }
}


lib/flutter_tableview.dart
library flutter_tableview;

import 'package:flutter/material.dart';

typedef int RowCountAtSection(int section);
typedef Widget ListViewFatherWidgetBuilder(
    BuildContext context, Widget canScrollWidget);
typedef Widget SectionHeaderBuilder(BuildContext context, int section);
typedef Widget CellBuilder(BuildContext context, int section, int row);
typedef double CellHeight(BuildContext context, int section, int row);
typedef double SectionHeaderHeight(BuildContext context, int section);

const String ErorrFlagBegin = '\n\n\n====================FlutterTableView  Error========================\n\n\n\n';

const String ErorrFlagEnd =   '\n\n\n\n\n==================================================================\n\n\n\n.';

class FlutterTableView extends StatefulWidget {

  FlutterTableView({
    @required this.sectionCount,
    @required this.rowCountAtSection,
    @required this.sectionHeaderBuilder,
    @required this.cellBuilder,
    @required this.sectionHeaderHeight,
    @required this.cellHeight,
    this.listViewFatherWidgetBuilder,
    this.controller,
    this.physics,
    this.shrinkWrap = false,
    this.padding = const EdgeInsets.all(0.0),
    this.cacheExtent = 30.0,
    this.backgroundColor = Colors.transparent,
  })  : assert(
  (sectionCount != null && sectionCount > 0),
  '$ErorrFlagBegin sectionCount must > 0 and could not be null. $ErorrFlagEnd',
  ),
        assert(
        (rowCountAtSection != null),
        '$ErorrFlagBegin function rowCountAtSection could not be null. $ErorrFlagEnd',
        ),
        assert(
        (sectionHeaderBuilder != null),
        '$ErorrFlagBegin function sectionHeaderBuilder could not be null. $ErorrFlagEnd',
        ),
        assert(
        (cellBuilder != null),
        '$ErorrFlagBegin function cellBuilder could not be null. $ErorrFlagEnd',
        ),
        assert(
        (sectionHeaderHeight != null),
        '$ErorrFlagBegin function sectionHeaderHeight could not be null. $ErorrFlagEnd',
        ),
        assert(
        (cellHeight != null),
        '$ErorrFlagBegin function cellHeight could not be null. $ErorrFlagEnd',
        );

  @override
  _FlutterTableViewState createState() => _FlutterTableViewState();

  /// How many section.
  final int sectionCount;

  /// How many item in each section.
  final RowCountAtSection rowCountAtSection;

  /// You can through sectionHeaderBuilder create section header widget.
  /// Each section has at most one headWidget.
  /// In a special section, if you don't need section header widget, you can return null.
  final SectionHeaderBuilder sectionHeaderBuilder;

  /// You can through cellBuilder create items.
  final CellBuilder cellBuilder;

  /// return each item widget height.
  final CellHeight cellHeight;

  /// return each section header widget height.
  final SectionHeaderHeight sectionHeaderHeight;

  /// You can wrap a widget for listView
  final ListViewFatherWidgetBuilder listViewFatherWidgetBuilder;

  /// see ScrollView controller
  final ScrollController controller;

  /// see ScrollView physics
  final ScrollPhysics physics;

  /// see ScrollView shrinkWrap
  final bool shrinkWrap;

  final EdgeInsetsGeometry padding;

  /// see ScrollView cacheExtent
  final double cacheExtent;

  final Color backgroundColor;
}

class _FlutterTableViewState extends State<FlutterTableView> {
  ////////////////////////////////////////////////////////////////////
  //                          variables
  ////////////////////////////////////////////////////////////////////
  SectionHeaderModel currentHeaderModel;
  int totalItemCount = 0;
  List<SectionHeaderModel> sectionHeaderList = List();
  List<int> sectionTotalWidgetCountList = List();
  ScrollController scrollController;
  ListView listView;
  Widget listViewFatherWidget;


  ////////////////////////////////////////////////////////////////////
  //                         init function
  ////////////////////////////////////////////////////////////////////
  void _initBaseData() {

    this.totalItemCount = 0;
    this.sectionHeaderList.clear();
    this.sectionTotalWidgetCountList.clear();

    double offsetY = 0;
    for (int section = 0; section < widget.sectionCount; section++) {
      int rowCount = widget.rowCountAtSection(section);
      Widget sectionHeader = widget.sectionHeaderBuilder(context, section);
      double sectionHeight;
      if (sectionHeader != null) {
        sectionHeight = this.widget.sectionHeaderHeight(context, section);
      } else {
        sectionHeight = 0;
      }

      double sectionHeaderY = offsetY;

      offsetY += sectionHeight;

      int sectionWidgetCount = sectionHeader == null ? rowCount : rowCount + 1;
      sectionTotalWidgetCountList.add(sectionWidgetCount);
      this.totalItemCount += sectionWidgetCount;

      for (int row = 0; row < rowCount; row++) {
        offsetY += this.widget.cellHeight(context, section, row);
      }

      SectionHeaderModel model = SectionHeaderModel(
        y: sectionHeaderY,
        sectionMaxY: offsetY,
        height: sectionHeight,
        section: section,
        headerWidget: sectionHeader,
      );
      sectionHeaderList.add(model);
    }
  }

  void _initScrollController() {

    this.scrollController = this.widget.controller;
    if(this.scrollController == null) {
      this.scrollController = ScrollController();
    }

    this.scrollController.addListener(() {
      double offsetY = this.scrollController.offset;

      if (offsetY <= 0.0) {
        this._updateCurrentSectionHeaderModel(null, 0);
      } else {
        int section = 0;
        for (int i = 0; i < this.sectionHeaderList.length; i++) {
          SectionHeaderModel model = this.sectionHeaderList[i];
          if (offsetY >= model.y && offsetY <= model.sectionMaxY) {
            section = i;
            break;
          }
        }

        SectionHeaderModel model = this.sectionHeaderList[section];
        double delta = model.sectionMaxY - this.scrollController.offset;
        double topOffset;
        if (delta >= model.height) {
          topOffset = 0.0;
        } else {
          topOffset = delta - model.height;
        }
        this._updateCurrentSectionHeaderModel(model, topOffset);
      }
    });
  }

  ////////////////////////////////////////////////////////////////////
  //                       create listView
  ////////////////////////////////////////////////////////////////////
  void _createListView() {
    if(this.listView == null) {
      this.listView = ListView.builder(
        controller: this.scrollController,
        physics: AlwaysScrollableScrollPhysics(),
        shrinkWrap: false,
        itemBuilder: (BuildContext context, int index) {
          Widget itemWidget;
          RowSectionModel model = this._getRowSectionModel(index);
          double height;
          if (model.row == 0 && model.haveHeaderWidget) {
            itemWidget = this.sectionHeaderList[model.section].headerWidget;
            height = this.widget.sectionHeaderHeight(context, model.section);
          } else {
            int row = model.haveHeaderWidget == false ? model.row : model.row - 1;
            itemWidget = this.widget.cellBuilder(context, model.section, row);
            height = this.widget.cellHeight(context, model.section, row);
          }

          return ConstrainedBox(
            constraints: BoxConstraints(
              minHeight: height,
              maxHeight: height,
            ),
            child: itemWidget,
          );
        },
        itemCount: this.totalItemCount,
      );
    }
  }


  ////////////////////////////////////////////////////////////////////
  //                      tool function
  ////////////////////////////////////////////////////////////////////
  RowSectionModel _getRowSectionModel(int index) {
    int passCount = 0;
    for (int section = 0;
    section < this.sectionTotalWidgetCountList.length;
    section++) {
      int currentSectionWidgetCount = this.sectionTotalWidgetCountList[section];
      if (index >= passCount && index < passCount + currentSectionWidgetCount) {
        int row = index - passCount;
        bool haveSectionHeaderWidget =
            this.sectionHeaderList[section].headerWidget != null;
        RowSectionModel model = RowSectionModel(
          section: section,
          row: row,
          haveHeaderWidget: haveSectionHeaderWidget,
        );

        return model;
      }

      passCount += currentSectionWidgetCount;
    }
    return null;
  }

  void _updateCurrentSectionHeaderModel(
      SectionHeaderModel model, double topOffset) {
    bool needSetState = false;
    if (model == null) {
      if (this.currentHeaderModel != null) {
        this.currentHeaderModel = null;
        needSetState = true;
      }
    } else if (this.currentHeaderModel == null) {
      this.currentHeaderModel = model;
      this.currentHeaderModel.topOffset = topOffset;
      needSetState = true;
    } else {
      if (model != this.currentHeaderModel) {
        this.currentHeaderModel = model;

        needSetState = true;
      } else if (model.topOffset != topOffset) {
        needSetState = true;
      }

      this.currentHeaderModel.topOffset = topOffset;
    }

    if (needSetState == true) {
      setState(() {});
    }
  }

  ////////////////////////////////////////////////////////////////////
  //                          life cycle
  ////////////////////////////////////////////////////////////////////

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    this.scrollController.dispose();
  }

  @override
  void initState() {
    this._initBaseData();
    this._initScrollController();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (this.listView == null) {
      this._createListView();
    }

    if(this.widget.listViewFatherWidgetBuilder != null && this.listViewFatherWidget == null) {
      this.listViewFatherWidget = this.widget.listViewFatherWidgetBuilder(context, this.listView);
    }

    Widget listViewWidget = this.listViewFatherWidget ?? this.listView;


    if (this.currentHeaderModel != null &&
        this.currentHeaderModel.headerWidget != null) {
      return Container(
        padding: this.widget.padding,
        color: widget.backgroundColor,
        child: Stack(
          children: <Widget>[
            Container(
              color: Colors.transparent,
              child: listViewWidget,
            ),
            Positioned(
              top: this.currentHeaderModel.topOffset,
              left: 0.0,
              right: 0.0,
              height: this.currentHeaderModel.height,
              child: this.currentHeaderModel.headerWidget,
            ),
          ],
        ),
      );
    }
    return Container(
      padding: this.widget.padding,
      color: this.widget.backgroundColor,
      child: Stack(
        children: [
          Container(
            color: Colors.transparent,
            child: listViewWidget,
          ),
        ],
      ),
    );
  }
}


////////////////////////////////////////////////////////////////////
//                          model class
////////////////////////////////////////////////////////////////////

class RowSectionModel {
  RowSectionModel({
    @required this.section,
    @required this.row,
    @required this.haveHeaderWidget,
  });

  final int section;
  final int row;
  final bool haveHeaderWidget;
}

class SectionHeaderModel {
  SectionHeaderModel({
    this.y,
    this.sectionMaxY,
    this.height,
    this.section,
    this.headerWidget,
  });

  final double y;
  final double sectionMaxY;
  final double height;
  final int section;
  final Widget headerWidget;

  double topOffset;
}


example/lib/demo_tool/route_tool.dart
import 'package:flutter/material.dart';

typedef PopCallback = Function(dynamic value);

Future<dynamic> push(BuildContext context, Widget page, {String routeName, PopCallback popCallBack}) {

  MaterialPageRoute route = MaterialPageRoute(builder: (BuildContext context){

    return page;
  });
  return Navigator.of(context).push(route).then((onValue){

    if(popCallBack != null) {

      popCallBack(onValue);
    }
  });
}

void pop(BuildContext context, {var callBackValue}) {

  if(callBackValue != null) {
    Navigator.of(context).pop(callBackValue);
  }
  else {
    Navigator.of(context).pop();
  }
}

void popToPage(BuildContext context, String pageClassName) {
  Navigator.of(context).popUntil(ModalRoute.withName(pageClassName));
}





double screenWidth;
double screenHeight;
double pxRatio;
MediaQueryData mediaQuery;

void initScreen(
    {@required BuildContext context}) {

  if(mediaQuery == null) {
    mediaQuery = MediaQuery.of(context);
    screenWidth = mediaQuery.size.width;
    screenHeight = mediaQuery.size.height;
  }
}

example/lib/demo_page/wrap_refresh_demo_page.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_tableview/flutter_tableview.dart';

//tool
import 'package:example/demo_tool/route_tool.dart';

//easyrefresh
import 'package:flutter_easyrefresh/easy_refresh.dart';
import 'package:flutter_easyrefresh/material_header.dart';
import 'package:flutter_easyrefresh/material_footer.dart';

class WrapRefreshDemoPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('wrap refresh demo'),
      ),
      body: WrapRefreshDemoPageBody(),
    );
  }
}

class WrapRefreshDemoPageBody extends StatefulWidget {
  @override
  _WrapRefreshDemoPageBodyState createState() =>
      _WrapRefreshDemoPageBodyState();
}

class _WrapRefreshDemoPageBodyState extends State<WrapRefreshDemoPageBody> {
  //key for flutter_easyrefresh
  //about flutter_easyrefresh : https://github.com/xuelongqy/flutter_easyrefresh
  GlobalKey<EasyRefreshState> _easyRefreshKey =
      new GlobalKey<EasyRefreshState>();
  GlobalKey<RefreshHeaderState> _headerKey =
      new GlobalKey<RefreshHeaderState>();
  GlobalKey<RefreshFooterState> _footerKey =
      new GlobalKey<RefreshFooterState>();

  //if dataSourceList data changed, must call setState(() {});
  List<List<String>> dataSourceList = [
    [
      'images/icon_00.png',
      'images/icon_01.png',
      'images/icon_02.png',
      'images/icon_03.png',
    ],
    [
      'images/icon_04.png',
      'images/icon_05.png',
      'images/icon_06.png',
      'images/icon_07.png',
      'images/icon_08.png',
    ],
  ];
  final double widthHeightRatio = 700.0 / 310.0;
  final double extraPaddingBottom = 30.0;

  //============
  //  functions
  //============
  int _rowCountAtSection(int section) {
    return this.dataSourceList[section].length;
  }

  Widget _sectionHeaderBuilder(BuildContext context, int section) {
    String message = section == 0 ? 'HOT SELL' : 'NEW';

    return InkWell(
      onTap: () {
        print('click section header. -> section:$section');
      },
      child: Container(
        color: Color.fromRGBO(220, 220, 220, 1),
        alignment: Alignment.center,
        height: 100,
        child: Text(
          '$message',
          style: TextStyle(color: Colors.brown, fontWeight: FontWeight.w700),
        ),
      ),
    );
  }

  Widget _cellBuilder(BuildContext context, int section, int row) {
    List<String> sectionDataList = this.dataSourceList[section];
    double paddingBottom =
        row != (sectionDataList.length - 1) ? 0 : this.extraPaddingBottom;

    return InkWell(
      onTap: () {
        print('click cell item. -> section:$section row:$row');
      },
      child: Container(
          padding: EdgeInsets.fromLTRB(30, 30, 30, paddingBottom),
          child: ClipRRect(
            borderRadius: BorderRadius.all(Radius.circular(10.0)),
            child: Image.asset(
              '${sectionDataList[row]}',
              fit: BoxFit.cover,
            ),
          )),
    );
  }

  double _sectionHeaderHeight(BuildContext context, int section) {
    return 60;
  }

  double _cellHeight(BuildContext context, int section, int row) {
    List<String> sectionDataList = this.dataSourceList[section];
    double cellHeight = screenWidth / this.widthHeightRatio;
    cellHeight = row != (sectionDataList.length - 1)
        ? cellHeight
        : cellHeight + this.extraPaddingBottom;
    return cellHeight;
  }

  Widget _listViewFatherWidgetBuilder(BuildContext context, Widget listView) {
    return EasyRefresh(
      key: _easyRefreshKey,
      limitScroll: true,
      refreshHeader: MaterialHeader(key: _headerKey),
      refreshFooter: MaterialFooter(key: _footerKey),
      onRefresh: () async {
        print('onrefresh');
        await Future.delayed(Duration(seconds: 2));
        this.dataSourceList = this.dataSourceList.reversed.toList();
        setState(() {});
      },
      loadMore: () async {
        print('loadMore');
        setState(() {});
      },
      child: listView,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: FlutterTableView(
        sectionCount: this.dataSourceList.length,
        rowCountAtSection: _rowCountAtSection,
        sectionHeaderBuilder: _sectionHeaderBuilder,
        cellBuilder: _cellBuilder,
        sectionHeaderHeight: _sectionHeaderHeight,
        cellHeight: _cellHeight,
        listViewFatherWidgetBuilder: _listViewFatherWidgetBuilder,
      ),
    );
  }
}


example/lib/demo_page/simple_demo_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_tableview/flutter_tableview.dart';

class SimpleDemoPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('simple demo'),
      ),
      body: SimpleDemoPageBody(),
    );
  }
}

class SimpleDemoPageBody extends StatefulWidget {
  @override
  _SimpleDemoPageBodyState createState() => _SimpleDemoPageBodyState();
}

class _SimpleDemoPageBodyState extends State<SimpleDemoPageBody> {
  // How many section.
  int sectionCount = 3;

  // Get row count.
  int _rowCountAtSection(int section) {
    if (section == 0) {
      return 5;
    } else if (section == 1) {
      return 10;
    } else {
      return 20;
    }
  }

  // Section header widget builder.
  Widget _sectionHeaderBuilder(BuildContext context, int section) {
    return InkWell(
      onTap: () {
        print('click section header. -> section:$section');
      },
      child: Container(
        alignment: Alignment.centerLeft,
        padding: EdgeInsets.only(left: 16.0),
        color: Color.fromRGBO(220, 220, 220, 1),
        height: 100,
        child: Text('I am section header -> section:$section'),
      ),
    );
  }

  // cell item widget builder.
  Widget _cellBuilder(BuildContext context, int section, int row) {
    return InkWell(
      onTap: () {
        print('click cell item. -> section:$section row:$row');
      },
      child: Container(
        padding: EdgeInsets.only(left: 16.0),
        alignment: Alignment.centerLeft,
        decoration: BoxDecoration(
            border: Border(
                bottom: BorderSide(
          color: Color.fromRGBO(240, 240, 240, 1),
        ))),
        height: 50.0,
        child: Text('I am cell -> section:$section  row$row'),
      ),
    );
  }

  // Each section header height;
  double _sectionHeaderHeight(BuildContext context, int section) {
    return 50.0;
  }

  // Each cell item widget height.
  double _cellHeight(BuildContext context, int section, int row) {
    return 50.0;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      //FlutterTableView
      child: FlutterTableView(
        sectionCount: sectionCount,
        rowCountAtSection: _rowCountAtSection,
        sectionHeaderBuilder: _sectionHeaderBuilder,
        cellBuilder: _cellBuilder,
        sectionHeaderHeight: _sectionHeaderHeight,
        cellHeight: _cellHeight,
      ),
    );
  }
}
